django/db/models/fields/related_descriptors.py

```diff
@@ -1130,6 +1130,41 @@ def get_prefetch_queryset(self, instances, queryset=None):
                 False,
             )
 
+        @property
+        def constrained_target(self):
+            # If the through relation's target field's foreign integrity is
+            # enforced, the query can be performed solely against the through
+            # table as the INNER JOIN'ing against target table is unnecessary.
+            if not self.target_field.db_constraint:
+                return None
+            db = router.db_for_read(self.through, instance=self.instance)
+            if not connections[db].features.supports_foreign_keys:
+                return None
+            hints = {'instance': self.instance}
+            manager = self.through._base_manager.db_manager(db, hints=hints)
+            filters = {self.source_field_name: self.instance.pk}
+            # Nullable target rows must be excluded as well as they would have
+            # been filtered out from an INNER JOIN.
+            if self.target_field.null:
+                filters['%s__isnull' % self.target_field_name] = False
+            return manager.filter(**filters)
+
+        def exists(self):
+            constrained_target = self.constrained_target
+            if constrained_target is not None and superclass is Manager and (
+                self.prefetch_cache_name is not None):
```

Comment:
`self.prefetch_cache_name is not None` is too aggressive as it will disable the optimization for any relationship that can be prefetched.

I'd suggest moving the logic from `get_queryset` that lookups `self.instance._prefetched_objects_cache[self.prefetch_cache_name]` in its own method and use it as a predicate instead.

Reply:
Thanks for your feedback, it's been really helpful! I've submitted an update improving this prefetch check in my latest PR. 

By the way, is there a way to check for linting errors before updating my PR? I want to submit PRs without them in the future. Thanks again for your help.

Reply:
> By the way, is there a way to check for linting errors before updating my PR? I want to submit PRs without them in the future. Thanks again for your help.

Yes, you can configure pre-commit to run the lint checks before committing, please see:

https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#pre-commit-checks

Reply:
> Yes, you can configure pre-commit to run the lint checks before committing, please see:

Thanks for the suggestion!